### PR TITLE
Ensure that gcc path is only added once to DLL search path

### PR DESCRIPTION
**Pull request by maresb**
## Description

This is an attempt to solve

```python

tests\distributions\test_transform.py:347: in check_vectortransform_elementwise_logp
    self.check_transform_elementwise_logp(model, vector_transform=True)
tests\distributions\test_transform.py:318: in check_transform_elementwise_logp
    test_array_untransf = x_val_untransf.eval({x_val_transf: test_array_transf})
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\graph\basic.py:616: in eval
    self._fn_cache[inputs] = function(inputs, self)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\compile\function\__init__.py:315: in function
    fn = pfunc(
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\compile\function\pfunc.py:469: in pfunc
    return orig_function(
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\compile\function\types.py:1762: in orig_function
    fn = m.create(defaults)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\compile\function\types.py:1654: in create
    _fn, _i, _o = self.linker.make_thunk(
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\basic.py:245: in make_thunk
    return self.make_all(
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\vm.py:1244: in make_all
    raise_with_op(fgraph, node)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\utils.py:531: in raise_with_op
    raise exc_value.with_traceback(exc_trace)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\vm.py:1235: in make_all
    node.op.make_thunk(node, storage_map, compute_map, [], impl=impl)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\op.py:119: in make_thunk
    return self.make_c_thunk(node, storage_map, compute_map, no_recycling)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\op.py:84: in make_c_thunk
    outputs = cl.make_thunk(
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\basic.py:1209: in make_thunk
    cthunk, module, in_storage, out_storage, error_storage = self.__compile__(
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\basic.py:1129: in __compile__
    thunk, module = self.cthunk_factory(
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\basic.py:1653: in cthunk_factory
    module = cache.module_from_key(key=key, lnk=self)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\cmodule.py:1231: in module_from_key
    module = lnk.compile_cmodule(location)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\basic.py:1552: in compile_cmodule
    module = c_compiler.compile_str(
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\cmodule.py:2652: in compile_str
    return dlimport(lib_filename)
C:\Users\runneradmin\miniconda3\envs\pymc-test\Lib\site-packages\pytensor\link\c\cmodule.py:325: in dlimport
    os.add_dll_directory(os.path.dirname(gcc_path))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

path = 'C:\\Users\\runneradmin\\miniconda3\\envs\\pymc-test\\Library\\mingw-w64\\bin'

>   ???
E   FileNotFoundError: [WinError 206] The filename or extension is too long: 'C:\\Users\\runneradmin\\miniconda3\\envs\\pymc-test\\Library\\mingw-w64\\bin'
```

This is a wild guess at the cause since I don't have Windows, and it's appearing only in the PyMC test suite. In any case I think this is unlikely to hurt.

## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by ricardoV94** [comment]
> This is an attempt to solve

The suspense is killing me

**Comment by maresb** [comment]
> The suspense is killing me
Look at the stack trace next. ;)

**Comment by lucianopaz** [comment]
I haven’t tested this but I personally don’t understand how this could fix the problem. Aren’t we bound to hit the same problem again if we the Windows runner has some weird and long path to gcc?
I googled a bit and found that Windows has a protocol to use more bits for the path https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry. Perhaps we could also use that prefix for the windows paths? Anyway, adding gcc only once seems perfectly fine and I like it much more than what we had before.

**Comment by maresb** [comment]
@lucianopaz, I'm not protecting against long paths, I'm protecting against the path being added over and over again.

**Comment by lucianopaz** [comment]
> @lucianopaz, I'm not protecting against long paths, I'm protecting against the path being added over and over again.

Yes, I understand that, but I don’t understand why doing that prevents the error you were seeing with long path names.

**Comment by maresb** [comment]
My hypothesis is simply that the error message is wrong. If you look at my stack trace, the path that triggered the error has quite an ordinary length. So I am hoping that when we encounter those absurdly long Conda prefix paths that we're only triggering the limit because we add it multiple times.

Like I wrote above, I'm not sure this will solve the problem, but I suspect it might.

File: pytensor/link/c/cmodule.py [hunk 1]
```diff
@@ -271,6 +272,14 @@ def _get_ext_suffix():
     return dist_suffix
 
 
+@cache
```
Comment (twiecki) [inline pytensor/link/c/cmodule.py:275]: Maybe add a comment of why we're caching this as it doesn't really make sense otherwise.

Comment (maresb) [reply pytensor/link/c/cmodule.py]: Great point, done!

**Review by twiecki — APPROVED** [review]

**Review by lucianopaz — APPROVED** [review]

**Comment by HarshvirSandhu** [comment]
I was working on #685 but when I run pre-commit locally, mypy fails with this message:
```
!!!!!!!!!
1 files unexpectedly failed.
pytensor/link/c/cmodule.py
These files did not fail before, so please check the above output for errors in {'pytensor/link/c/cmodule.py'} and fix them.
```

**Comment by maresb** [comment]
Hi @HarshvirSandhu, thanks for letting us know, but could you please provide more details? In particular, could you please provide the specific mypy error that is reported for `pytensor/link/c/cmodule.py`?

This is a bit strange since I'm unable to reproduce it locally, and usually these things are caught by the automated tests.

**Comment by Dhruvanshu-Joshi** [comment]
Hello @maresb and @HarshvirSandhu ,
I am facing this same error when mypy is run locally:
```
[pytensor\link\c\cmodule.py]
pytensor\link\c\cmodule.py:287: error: Unused "type: ignore" comment
```
https://github.com/pymc-devs/pytensor/blob/bc3dda0ef9ea9d20934928ad0ef98ecd70eb705b/pytensor/link/c/cmodule.py#L287

**Comment by maresb** [comment]
Thanks @Dhruvanshu-Joshi for the info! It seems quite mysterious why we are now seeing different results when mypy is run locally vs in the CI. (We see that the CI checks are failing in #702 after you added `cmodule.py` to `scripts/mypy-failing.txt` because there are no errors in the CI.) Are you running mypy via pre-commit? What operating system are you using?

**Comment by Dhruvanshu-Joshi** [comment]
@maresb I am running mypy via pre-commit. Currently, I am using windows 11.

**Pull request merged**